net: NAD reference live-update STP (#74)
RedHatQE/openshift-virtualization-tests-design-docs: stps/sig-network/hotpluggable-nad-ref.md

    - Scalability: no new scale requirements.
    - UI: no UI changes introduced by this feature.

#### **2. Known Limitations**

- **Non-migratable VMs:** Changing the secondary network on non-migratable VMs is not supported; the VM will not reconnect to the new network without a restart.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Binding support:** Any supported bridge binding only.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Guest network config:** Feature does not change guest-side configuration; guest may need separate reconfiguration for new network (e.g., IP).
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Single-node clusters (SNO):** Cannot be supported because migration is not possible.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Migrating between CNI types:** Explicit non-goal.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Changing network binding/plugin:** Explicit non-goal.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Seamless network connectivity until action is completed:** Not guaranteed by design.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **No limitation of migration retries because of missing Network Attachment Definition:** Explicit non-goal.

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Author: Those are known limitations
Out of scope refer to scenarios we don't test

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer: Sorry @azhivovk I still don't understand this sentence. Perhaps you mean to assume that the target NAD exists?

#### **3. Technology and Design Review**

- [x] **Developer Handoff/QE Kickoff**
  - *Key takeaways and concerns:* Meeting where Dev introduced QE through NAD change in VM, migration condition handling, and compatibility with DNC.

- [x] **Technology Challenges**
  - *List identified challenges:* The feature relies on live migration, which is not supported on all cluster topologies (e.g., single-node clusters), limiting the environments in which this feature can be tested.
  - *Impact on testing approach:* Tests must run on multi-node clusters where live migration is available.

- [x] **API Extensions**
  - *List new or modified APIs:* No new APIs introduced by this feature.
  - *Testing impact:* N/A

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*

- [x] **Topology Considerations**
  - *Describe topology requirements:* No known limitations when considering different topologies.

>   - *Impact on test design:* No specific impact on test design beyond standard live migration requirements.

Author: This section is about affected topologies by the feature, no?

>   - *Impact on test design:* No specific impact on test design beyond standard live migration requirements.

Author: The feature is not impacting on topology, it can be tested on multi node clusters

>   - *Impact on test design:* No specific impact on test design beyond standard live migration requirements.

Author: It is mentioned in test environment and known limitations

  - *Note any gaps or missing criteria:* None

- [x] **Non-Functional Requirements (NFRs)**
  - *List applicable NFRs and their targets:* Documentation update required.
  - *Note any NFRs not covered and why:*
    - Performance: no new performance requirements introduced.
    - Security: no new RBAC or authentication changes.
    - Monitoring/Observability: no new metrics or alerts required. The existing VM network interface info metric should be validated to confirm it reflects the updated network attachment after the swap.
    - Scalability: the feature has scale implications — large-scale NAD reference updates are throttled by the cluster's migration parallelism limits. This is accepted as a known risk rather than a test requirement (see Risks, Section II.5). Documentation should reflect this limitation.
    - UI: no new UI components introduced, but existing VM network display must be validated by the UI team (see Usability Testing in Section II.2).

#### **2. Known Limitations**

- **Non-migratable VMs:** Changing the secondary network on non-migratable VMs is not supported; the VM will not reconnect to the new network without a restart.

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

- **Binding support:** Secondary network reassignment is supported for bridge binding only; other binding types are not in scope for this feature.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Guest network config:** Feature does not change guest-side configuration; guest may need separate reconfiguration for new network (e.g. IP).

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

- **Single-node clusters (SNO):** Cannot be supported because migration is not possible.

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

> - **Migrating between CNI types:** Explicit non-goal.

Reviewer 3: **LOW — I.2: “Explicit non-goal” phrasing reads as a scope decision, not a product constraint**

Per `AGENTS.md`: *“Limitations are constraints (not scope decisions).”* The phrase “Explicit non-goal” reads as a deliberate scoping choice by QE, not a confirmed product constraint. Applies to this entry and the two that follow (Changing network binding/plugin; Limiting migration retries on failure).

**Proposed fix (same pattern for all three):**

- **Migrating between CNI types:** Not supported; the feature is scoped to reassigning secondary networks within the same CNI type only.
- **Changing network binding/plugin:** Not supported; changing the binding type or plugin as part of a network reassignment is not in scope for this feature.
- **Limiting migration retries on failure:** Not supported; migration retry limits are not configurable as part of this feature.

---
*Reviewed by Claude Sonnet 4.6*

Author: Rephrased it slightly different since we can understand the limitation by the entry name

  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Changing network binding/plugin:** Explicit non-goal.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Seamless network connectivity until action is completed:** Not guaranteed by design. After the NAD update completes, connectivity is expected to be preserved without guest intervention, except when the new network requires explicit guest-side reconfiguration (e.g. IP change).

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

- **Limiting migration retries on failure:** Explicit non-goal.
  - *Sign-off:* Ronen Sde-Or, 04/2026

#### **3. Technology and Design Review**

- [x] **Developer Handoff/QE Kickoff**

>   - *Key takeaways and concerns:* Meeting where Dev introduced QE through NAD change in VM and migration condition handling.

Reviewer 3: **HIGH — I.3: Developer Handoff description lacks substance**

The current text is a single generic sentence that captures no actionable QE takeaways:

> *“Meeting where Dev introduced QE through NAD change in VM and migration condition handling.”*

Per `AGENTS.md`: *“Developer Handoff is a single generic sentence with no substance”* is an explicit common rejection reason. The field must describe what was actually discussed — design decisions that affect test planning, edge cases flagged by Dev, concerns raised by QE, and any constraints surfaced during the meeting.

**Proposed text:**

- [x] **Developer Handoff/QE Kickoff**
  - *Key takeaways and concerns:* Dev walked QE through the NAD swap flow: the VM is reconnected via live migration, which preserves the guest interface configuration by design. If migration fails mid-swap, the VM remains on the original network — no manual rollback is needed. QE concern raised: timing the “no connectivity on old network” assertion during the migration window may produce false positives; Dev confirmed a stable state is reached once migration completes and that is the right point to assert.

Replace with the actual notes from your kickoff meeting.

---
*Reviewed by Claude Sonnet 4.6*

Author: Rephrased and added

- [x] **Technology Challenges**
  - *List identified challenges:* The feature relies on live migration, which is not supported on all cluster topologies (e.g. single-node clusters), limiting the environments in which this feature can be tested.
  - *Impact on testing approach:* Tests must run on multi-node clusters where live migration is available.

- [x] **API Extensions**
  - *List new or modified APIs:* No new APIs introduced by this feature.
  - *Testing impact:* N/A

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*

- [x] **Topology Considerations**

>   - *Describe topology requirements:* No known limitations when considering different topologies.

Reviewer 3: **MEDIUM — I.3: Topology Considerations contradicts Technology Challenges**

Technology Challenges states *“Tests must run on multi-node clusters where live migration is available”*, but Topology Considerations says *“No known limitations when considering different topologies”* — a direct contradiction. The follow-up *“No specific impact on test design beyond standard live migration requirements”* is also vague: those requirements are never defined.

**Proposed text:**

- [x] **Topology Considerations**
  - *Describe topology requirements:* Tests require a multi-node cluster with live migration available (minimum 2 worker nodes). SNO is excluded — the feature does not support it (see Known Limitations, Section I.2).
  - *Impact on test design:* All test scenarios must run on multi-node clusters; SNO environments are not valid test targets.

---
*Reviewed by Claude Sonnet 4.6*

Author: Done

  - *Impact on test design:* No specific impact on test design beyond standard live migration requirements.

### **II. Software Test Plan (STP)**

#### **1. Scope of Testing**

**Testing Goals**

- **[P0]** Verify that a running VM with bridge binding on a Linux bridge CNI secondary network can be reconnected to a different secondary network without rebooting, has connectivity on the new network, and is no longer reachable on the previous network.
- **[P1]** Verify that a running VM connected to a localnet secondary network can be reconnected to a different localnet network without rebooting and has connectivity on the new network.

> - **[P1]** Verify that when NAD references for multiple secondary network interfaces on the same running VM are updated at once, all interfaces establish connectivity on their respective new networks.

Reviewer 3: **MEDIUM — II.1: Testing goal missing configuration dimensions**

Per `AGENTS.md`: *“Testing goals are fully actionable: they name all configuration dimensions needed to implement the test (e.g., for networking: both the binding type and the CNI in use).”* This goal specifies neither binding type nor CNI, leaving test authors to guess.

**Proposed text:**

- **[P1]** Verify that when NAD references for multiple secondary network interfaces (bridge binding, Linux bridge CNI) on the same running VM are updated at once, all interfaces establish connectivity on their respective new networks.

---
*Reviewed by Claude Sonnet 4.6*

Author: Rephrased differently

- **[P1]** Verify that attempting to change the secondary network VLAN on a non-migratable VM with bridge binding on a Linux bridge CNI secondary network does not succeed, and the VM remains on the original network.

> - **[P1]** Verify that the VM network interface monitoring data is updated to reflect the new network attachment after a secondary network reassignment.

Reviewer 3: **MEDIUM — II.1: Testing goal missing configuration dimensions**

Per `AGENTS.md`: *“Testing goals are fully actionable: they name all configuration dimensions needed to implement the test (e.g., for networking: both the binding type and the CNI in use).”* This goal specifies neither binding type nor CNI.

**Proposed text:**

- **[P1]** Verify that after a secondary network reassignment (bridge binding, Linux bridge CNI), the VM network interface monitoring data is updated to reflect the new network attachment.

---
*Reviewed by Claude Sonnet 4.6*

Author: Rephrased

- [ ] **Performance Testing** — Validates feature performance meets requirements (latency, throughput, resource usage)
  - *Details:* Not applicable; no new performance requirements introduced for this feature.

- [ ] **Scale Testing** — Validates feature behavior under increased load and at production-like scale
  - *Details:* Not applicable; no new scale requirements introduced for this feature.

- [ ] **Security Testing** — Verifies security requirements, RBAC, authentication, authorization
  - *Details:* Not applicable; no new RBAC or authentication changes introduced.

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g. status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g. status conditions, events, or notifications indicating success or failure)

>   - *Details:* No new UI components introduced. The UI team should validate that the VM network display correctly reflects the updated NAD reference after the swap completes. Displaying the previous IP is expected behavior.

Reviewer 3: **HIGH — II.2: Usability Testing checkbox checked but details don’t describe what QE tests**

The details say *“The UI team should validate…”* but per `AGENTS.md`, if Usability is checked, the content must describe what QE tests — not delegate it. Since the QE UI team owns this, describe the concrete verification they will perform.

**Proposed text:**

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - *Details:* QE UI team will validate that the VM network display correctly reflects the updated NAD reference after the swap completes, and that the admin receives clear observable feedback (e.g. status condition or event) indicating when the network swap has completed successfully.

---
*Reviewed by Claude Sonnet 4.6*

Reviewer: This ^ + I don't understand `Displaying the previous IP is expected behavior.`  How is L3 related?

Author: We don't have clear indication of swap completed. 
In UI we expect to see change of network name and IP preserved.

- [ ] **Monitoring** — Does the feature require metrics and/or alerts?

>   - *Details:* Not applicable; no new metrics or alerts required.

Reviewer 3: **HIGH — II.2: Monitoring unchecked but actively tested**

The Monitoring checkbox is unchecked with *“Not applicable; no new metrics or alerts required”*, yet there is a **P1 testing goal** and a matching **P1 test scenario** in Section III that explicitly test monitoring behavior. The checkbox must be checked and the details updated.

**Proposed text:**

- [x] **Monitoring** — Does the feature require metrics and/or alerts?
  - *Details:* No new metrics or alerts are introduced. The existing VM network interface info metric will be validated to confirm it correctly reflects the updated network attachment after the swap completes (see P1 testing goal and Section III scenario).

---
*Reviewed by Claude Sonnet 4.6*

Author: Done

- **Network:** OVN-Kubernetes, IPv4, IPv6, secondary bridge network, localnet

- **Required Operators:** NMState Operator (for bridge interface configuration)

- **Platform:** Bare metal, PSI

- **Special Configurations:** N/A

#### **3.1. Testing Tools & Frameworks**

- **Test Framework:** Standard

- **CI/CD:** N/A

- **Other Tools:** N/A

#### **4. Entry Criteria**

The following conditions must be met before testing can begin:

- [x] Requirements and design documents are **approved and merged**
- [x] Test environment can be **set up and configured** (see Section II.3 - Test Environment)

#### **5. Risks**

**Timeline/Schedule**

- **Risk:** None identified.
  - **Mitigation:** N/A
  - *Estimated impact on schedule:* N/A

**Test Coverage**

- **Risk:** Large-scale NAD reference updates are not tested; changing NADs on many VMs simultaneously is throttled by the cluster's migration parallelism limits and may take time to stabilize.
  - **Mitigation:** Accepted as out of scope for this release. Feature documentation should reflect this scale limitation.
  - *Areas with reduced coverage:* Scale behavior of concurrent NAD reference updates.

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

  - **Mitigation:** Accepted per feature scope - bridge binding is the only supported binding type for this feature.
  - *Areas with reduced coverage:* Other network binding types.
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Test Environment**

- **Risk:** None.
  - **Mitigation:** Cluster topology requirements are documented in Section II.3.
  - *Missing resources or infrastructure:* None

**Untestable Aspects**

- **Risk:** None — all components are testable.
  - **Mitigation:** Connection continuity during the swap is a non-goal by design and is documented in Known Limitations (Section I.2).
  - *Alternative validation approach:* N/A

**Resource Constraints**

- **Risk:** No special resource requirements.
  - **Mitigation:** New tests will run on network existing lanes.
  - *Current capacity gaps:* None

**Dependencies**

- **Risk:** No blocking external dependencies identified.
  - **Mitigation:** N/A
  - *Dependent teams or components:* None

---

### **III. Test Scenarios & Traceability**

- **[VIRTSTRAT-560]** — As a VM owner, I want to change my VM's VLAN without rebooting so that my workload remains available during network reconfiguration.

>   - *Test Scenario:* [Tier 2] Verify that a running VM with bridge binding on a Linux bridge CNI secondary network remains running after a VLAN change, establishes connectivity on the new VLAN, and is no longer reachable on the previous VLAN.

Reviewer 3: **MEDIUM — III: Guest interface preservation acceptance criterion has no traceable scenario**

The acceptance criterion from Section I.1 — *“The guest interface and its (IP) configuration is preserved during the transition”* — is not covered by any test scenario. Per `AGENTS.md`: *“Every Acceptance Criterion from Section I.1 is traceable to a scenario.”*

**Proposed fix — extend the P0 scenario:**

- *Test Scenario:* [Tier 2] Verify that a running VM with bridge binding on a Linux bridge CNI secondary network remains running after a VLAN change, establishes connectivity on the new VLAN, is no longer reachable on the previous VLAN, and the guest interface and its configuration are preserved throughout the transition without guest-side intervention.

---
*Reviewed by Claude Sonnet 4.6*

Author: Done

# Openshift-virtualization-tests Test plan

## **Live Update NetworkAttachmentDefinition Reference  - Quality Engineering Plan**

### **Metadata & Tracking**

- **Enhancement(s):** https://github.com/kubevirt/enhancements/blob/main/veps/sig-network/hotpluggable-nad-ref.md
- **Feature Tracking:** https://issues.redhat.com/browse/VIRTSTRAT-560
- **Epic Tracking:** https://redhat.atlassian.net/browse/CNV-72329
- **QE Owner(s):** Asia Khromov (@azhivovk)
- **Owning SIG:** sig-network
- **Participating SIGs:** sig-network

**Document Conventions (if applicable):**
NAD = NetworkAttachmentDefinition, a Multus secondary network definition

> DNC = Dynamic Networks Controller (Multus dynamic networks controller)

Reviewer 4: It looks like the 2 lines are supposed to be separated, but in preview it turns out
```
Document Conventions (if applicable): NAD = NetworkAttachmentDefinition, a Multus secondary network definition DNC = Dynamic Networks Controller (Multus dynamic networks controller)
```
Please search how to do that in markdown (<br> tag, double-space at the line-end, etc.).
Also note that this occurs in other places in the doc as well.

Author: Fixed

> - DNC = Dynamic Networks Controller (Multus dynamic networks controller)

Reviewer 5: DNC (Dynamic Networks Controller) is not productized and is not deployed downstream - please consider drooping it.

Author: Done

### **Feature Overview**

> This feature enables VM owners to change the secondary network of a running VM without rebooting. When the user updates the network reference in the VM configuration, the VM is reconnected to the new network (e.g., a different VLAN) without requiring a restart, preserving guest interface identity.

Reviewer 4: This overview doesn't mention the required migration.
On one hand it makes sense, because the migration part is an implementation detail; but OTOH, you do mention the required migration in other places in the STP (in the "Known Limitations", for example), so maybe it makes sense to mention it here, for example

```suggestion
..., the VM is reconnected to the new network (e.g., a different VLAN) without requiring a restart (but rather migrating the VM), preserving guest interface identity.
sdfd
```
@frenzyfriday WDYT?

Reviewer 6: yes, it makes sense to add the migration here. Though it is an implementation detail, the user right now has no other way to monitor the change than to check for the migration criteria

Author: You are right, it's important to mention it in the overview since we mentioned the migration limitations

> The supported scope is bridge binding on live-migratable VMs.

Reviewer 4: new-line issue

Author: Fixed

The supported scope is bridge binding on live-migratable VMs.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**
  - *List the key D/S requirements reviewed:* Change the secondary network of a running VM without requiring a restart and without any change in the guest interfaces.

- [x] **Understand Value and Customer Use Cases**
  - *Describe the feature's value to customers:* VM owners can re-assign new networks (e.g. a different VLAN) to VMs without reboot, avoiding workload impact and preserving guest interface identity (e.g. MAC).
  - *List the customer use cases identified:*

>     - As a VM owner, I want to re-assign my VM to a different VLAN without rebooting so that my VM connects to the new network without requiring a restart.

Reviewer 5: Should we also mention "adjustments to guest network configuration"?

Author: Done

- [x] **Testability**
  - *Note any requirements that are unclear or untestable:* Core flow is testable (see III. Test Scenarios & Traceability).

- [x] **Acceptance Criteria**
  - *List the acceptance criteria:*
    - The VM remains running (not rebooted) after the NAD reference update
    - The VM establishes connectivity on the new secondary network
    - The VM no longer has connectivity on the previous secondary network

>     - The guest interface identity (e.g., MAC address) is preserved

Reviewer 4: Redundant  comma after "e.g.".

Author: Done

  - *Note any gaps or missing criteria:* None

- [x] **Non-Functional Requirements (NFRs)**
  - *List applicable NFRs and their targets:* Documentation update required.
  - *Note any NFRs not covered and why:*
    - Performance: no new performance requirements introduced.
    - Security: no new RBAC or authentication changes.
    - Monitoring/Observability: no new metrics or alerts required.
    - Scalability: no new scale requirements.

>     - UI: no UI changes introduced by this feature.

Reviewer 6: Yes, as of now. i will talk to the UI team to get their opinions.

Reviewer: @frenzyfriday
Do you mean that the NAD field is currently editable in a running VM? Does the user have to edit the yaml through GUI? Does it work?

Reviewer 3: The STP author and feature lead needs to evaluate this from a customer perspective.
If you think it is not needed, then reason for it. Mentioning that no UI changes are introduced is not saying that it is not needed.

While discussing this with the UI team is nice, it does not answer the basic question: Is it needed from a customer perspective?
Usually such questions are answered by stakeholders, like the PM and/or UX.

Author: I asked for UI review, might need to add a new dedicated test

>     - UI: no new UI components introduced, but existing VM network display must be validated by the UI team (see Usability Testing in Section II.2).

Reviewer 5: Unfortunately, its not straight forward for a VM owner to understand whether the VM has completed the transition or not.

Author: The meaning of this is UI will add a new test to make sure the IP stays the same as before the NAD change, validate the NAD name and VLAN change

  - Does the feature maintain backward compatibility with previous API versions and configurations?
  - *Details:* Not applicable; no new API versions or configurations introduced.

- [x] **Upgrade Testing** — Validates upgrade paths from previous versions, data migration, and configuration preservation
  - *Details:* Verify that VMs with updated NAD references retain their network configuration and connectivity after cluster upgrade.

- [ ] **Dependencies** — Blocked by deliverables from other components/products. Identify what we need from other teams before we can test.
  - *Details:* Not applicable; no blocking external dependencies identified.

- [ ] **Cross Integrations** — Does the feature affect other features or require testing by other teams? Identify the impact we cause.
  - *Details:* Not applicable; no impact on other features or teams identified.

**Infrastructure**

- [ ] **Cloud Testing** — Does the feature require multi-cloud platform testing?
  - *Details:* Not applicable; feature requires bare-metal cluster with live migration support.

#### **3. Test Environment**

> 

Reviewer 4: If the feature is limited only to bridge-binding, as specified in the "Test Coverage" section below, then I think that multi-NIC nodes should also be mentioned here as a test environment requirement.

Author: I mentioned it in
`- **Network:** OVN-Kubernetes, IPv4, IPv6, secondary bridge network`

Do I need to mention it in more places? Cluster topology maybe?

Reviewer 4: My opinion is that you should explicitly mention that multi NIC nodes are required, either in network or cluster topology.
However, maybe you think that this mention is clear and enough, and in that case - I won't insist.

Author: Added to Cluster Topology

> - **Cluster Topology:** 3-master/3-worker bare-metal (minimum 2 workers)

Reviewer 4: Must it be bare-metal? Why?

Author: IIUC, it's the default cluster setup we mention in the STP

Reviewer 4: @rnetser is this true?
And if so - why? We shouldn't be limited only to BM IMO.

Author: Added PSI

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization version 4.22

> - **CPU Virtualization:** VT-x (Intel) or AMD-V enabled

Reviewer 4: So it won't be supported on ARM-based nodes?

Author: IIUC it's the equivalent of agnostic, as written in the [stp template](https://github.com/RedHatQE/openshift-virtualization-tests-design-docs/blob/main/stps/stp-template/stp.md): 

```
- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled
  <!-- Change only if specific CPU requirements exist -->
```

Reviewer 4: I think that the template's default corresponds to the current RHEL nodes we have, which use AMD.
Now that multi-arch is about to be supported, I think that this section should also mention ARM.

Author: Added ARM, thanks

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM

- **Special Hardware:** N/A

- **Storage:** ocs-storagecluster-ceph-rbd-virtualization

- **Network:** OVN-Kubernetes, IPv4, IPv6, secondary bridge network

- **Required Operators:** NMState Operator (for bridge interface configuration)

> - **Platform:** Bare metal

Reviewer 4: Same - why necessarily BM?

Author: Similar to https://github.com/RedHatQE/openshift-virtualization-tests-design-docs/pull/74#discussion_r3086828289

Reviewer: Can you please elaborate?

Author: BM is the default platform according to the [template](https://github.com/RedHatQE/openshift-virtualization-tests-design-docs/blob/main/stps/stp-template/stp.md?plain=1#L316) 
I added PSI

#### **5. Risks**

**Timeline/Schedule**

- [x] **Risk:** None identified.
  - **Mitigation:** N/A
  - *Estimated impact on schedule:* N/A
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Test Coverage**

- [x] **Risk:** Coverage is limited to bridge binding only; other binding types are not tested.
  - **Mitigation:** Accepted per feature scope - bridge binding is the only supported binding type for this feature.
  - *Areas with reduced coverage:* Other network binding types.
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Test Environment**

- [x] **Risk:** Live migration requires at least 2 worker nodes; SNO clusters cannot be used.
  - **Mitigation:** Ensure test environment has a minimum of 2 worker nodes as specified in Section II.3.
  - *Missing resources or infrastructure:* None identified at this time.
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Untestable Aspects**

- [x] **Risk:** Seamless network connectivity during the NAD swap is not guaranteed by design and cannot be tested as a pass/fail condition.
  - **Mitigation:** Tests will verify connectivity before and after the swap; transient disruption during swap is accepted.

>   - *Alternative validation approach:* Verify connectivity is restored on new network after swap completes.

Reviewer 4: Please be clear here, on whether the connectivity is expected to be seamlessly restored (meaning - a connection that was established before the NAD swap will disrupted during the change, and will eventually be restored without the user having to actively restore it), or rather the user will have to actively renew this connection.
If I understand the feature and its VEP correctly, then it's the second option.

Author: IIUC according to the VEP, the connectivity restoration is not seamlessly maintained and if the connectivity was not preserved after the NAD update, the user needs to manually renew the connections.
I added it here, thanks

Reviewer 6: I dont think the user needs to take manual steps after the process completes (if they change between CNIs that is a different scenario) - the connectivity should be there just after the migration completes.
Are there scenarios where the user may need to change IPs (or something similar) because of the new network?

Author: We expect the IP to stay the same but IIUC by the VEP, an active connection might be interrupted and must be renewed if so, or am I wrong and we expect any active connection to stay such?

Reviewer 6: I think you are referring to the non goal `Changing the guest's network configuration in case it have to change following the underlying network change.` ?
I meant that during the migration, connectivity may be dropped for some time but then it will be restored as the new pod is created (and the CNI connects it to the new network based on its network annotation) and the VM is unpaused. In the usual scenarios the user wont have to touch the guest to restore connectivity.
I added this non goal for scenarios where the VM will explicitly need to change its IP (or some other iface property) to connect to the new network.

Author: Got it.
I edited the risk accordingly, thanks

  - *Sign-off:* Ronen Sde-Or, 04/2026

**Resource Constraints**

- [x] **Risk:** No special resource requirements.
  - **Mitigation:** N/A
  - *Current capacity gaps:* None
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Dependencies**

- [x] **Risk:** No blocking external dependencies identified.
  - **Mitigation:** N/A
  - *Dependent teams or components:* None
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Other**

- [x] **Risk:** None identified.
  - **Mitigation:** N/A
  - *Sign-off:* Ronen Sde-Or, 04/2026

---

> ### **III. Test Scenarios & Traceability**

Reviewer 6: Should we add a Ipv6 scenario somewhere? And should we add a test for VLAN change explicitly?

Author: IPv6 is covered as a subtest within the existing scenarios,  not a separate user story.                                                                                 
I plan to follow the way we create new tests:  setup the VM networks according to existing cluster network stack - if we run on dual-stack cluster we test both IP families and if we run on single-stack cluster, we test accordingly.
Regarding the VLAN change - I edited the first test scenario

- **[VIRTSTRAT-560]** — As a VM owner, I want to re-assign my VM to a different secondary network without rebooting so that my workload remains available during network reconfiguration.
  - *Test Scenario:* [Tier 2] Verify that a running VM with bridge binding (OVN-Kubernetes) can be reconnected to a different secondary network without rebooting, has connectivity on the new network, and is no longer reachable on the previous network.
  - *Priority:* P0

- **[VIRTSTRAT-560]** — As a VM owner, I want to update multiple secondary networks on the same running VM and have connectivity on all of them.
  - *Test Scenario:* [Tier 2] Verify that multiple secondary networks on the same running VM can each be independently updated to different NADs and remain reachable.
  - *Priority:* P1

- **[VIRTSTRAT-560]** — As a VM owner, I want my VM to be live-migratable after a NAD reference update.
  - *Test Scenario:* [Tier 2] Verify that a VM can be successfully live-migrated after completing a NAD reference update, and maintains connectivity on the new network.
  - *Priority:* P2

> - **[VIRTSTRAT-560]** — As a VM owner, I want my VM to remain connected to its updated secondary network after a cluster upgrade.

Reviewer 6: The implementation did not add a breaking change. So is the cluster upgrade test necessary still?

Author: To me it sounds like a scenario we should test, but maybe I'm wrong
@EdDev WDYT?

Reviewer 3: I commented about it in the testing goals. I do not see any reason to have this, but if you can explain how things can get wrong, please fight for it.

Author: Removed, thanks

### **Metadata & Tracking**

- **Enhancement(s):** https://github.com/kubevirt/enhancements/blob/main/veps/sig-network/hotpluggable-nad-ref.md
- **Feature Tracking:** https://issues.redhat.com/browse/VIRTSTRAT-560
- **Epic Tracking:** https://redhat.atlassian.net/browse/CNV-72329
- **QE Owner(s):** Asia Khromov (@azhivovk)
- **Owning SIG:** sig-network
- **Participating SIGs:** sig-network

**Document Conventions (if applicable):**
- NAD = NetworkAttachmentDefinition, a Multus secondary network definition
- DNC = Dynamic Networks Controller (Multus dynamic networks controller)

### **Feature Overview**

This feature enables VM owners to change the secondary network of a running VM without rebooting. When the user updates the network reference in the VM configuration, the VM is reconnected to the new network (e.g. a different VLAN) without requiring a restart (but rather seamlessly migrating the VM), preserving guest interface identity.

The supported scope is bridge binding on live-migratable VMs.

---

### **I. Motivation and Requirements Review (QE Review Guidelines)**

#### **1. Requirement & User Story Review Checklist**

- [x] **Review Requirements**

>   - *List the key D/S requirements reviewed:* Change bridge binding NAD reference on a running VM without requiring a restart; the VM should reconnect to the new network seamlessly.

Reviewer 3: > ; the VM should reconnect to the new network seamlessly.

If you change the network, what does this sentence mean exactly?

Author: Rephrased

- [x] **Understand Value and Customer Use Cases**

>   - *Describe the feature's value to customers:* VM owners can re-assign new networks (e.g. a different VLAN) to VMs without reboot, avoiding workload impact and preserving guest interface identity (e.g. MAC).

Reviewer 3: > preserving guest interface identity (e.g. MAC)

I do not fully understand that, because the need is to never unplug it, not about keeping identity. In practice, I can unplug and plug back an interface and preserve its identity (that is the spec definition).

Author: Rephrased

Reviewer 3: I see no change.

Author: Now the change is visible

  - *List the customer use cases identified:*

>     - As a VM owner, I want to re-assign my VM to a different VLAN without rebooting so that my workload remains available during network reconfiguration.

Reviewer: IMHO the objective is not that the network "remains available" but rather that the network is actually changed.

Author: Done

- [x] **Testability**
  - *Note any requirements that are unclear or untestable:* Core flow is testable (see III. Test Scenarios & Traceability).

- [x] **Acceptance Criteria**

>   - *List the acceptance criteria:*

Reviewer: Please consider indicating the NAD reference update includes a change of VLAN. The reason for this comment is that a NAD is a technical entity whereas the VLAN change is what the customer is requiring to perform.  Therefore just indicating that a NAD swap had succeeded does not indicate that the feature is working.

Author: Added, thanks

    - The VM remains running (not rebooted) after changing the secondary network (e.g. VLAN change)
    - The VM establishes connectivity on the new secondary network (new VLAN)
    - The VM no longer has connectivity on the previous secondary network (previous VLAN)

>     - The guest interface identity (e.g. MAC address) is preserved

Reviewer 3: We can preserve identity but still fail: Unplug and plug again.
If you want to define a criteria, then I guess something like this is more accurate:
> The guest interface and its (IP) configuration is preserved during the transition.

Author: Done

  - *Note any gaps or missing criteria:* None

- [x] **Non-Functional Requirements (NFRs)**
  - *List applicable NFRs and their targets:* Documentation update required.
  - *Note any NFRs not covered and why:*
    - Performance: no new performance requirements introduced.
    - Security: no new RBAC or authentication changes.

>     - Monitoring/Observability: no new metrics or alerts required.

Reviewer 3: Will the reporting be updated automatically when moving between NADs? I remember that this information is collected.

Author: I asked for metrics review

Reviewer 7: I think adding a another test case for the metric kubevirt_vm_vnic_info verifying the value of the metric for the NAD is updated when swiching it.

Reviewer 3: @azhivovk , nothing was changed here. Please clarify.

Author: Now the change is visible

>     - Scalability: no new scale requirements.

Reviewer 3: I actually think this is effected, due to the mechanism the feature is using.
The use of live-migration has scale implications due to the limits imposed on the cluster level (on the amount of VMs allowed to perform live migration in parallel). Changes in large scale, will require time to stabilize and complete.

Author: Good point, how do we test it? Do we add this kind of test or other team?

Reviewer 3: We have two options:
- Consider this very important and opening a ticket to cover this at scale.
- Considering the factor and accepting it as a risk (out of scope?).

I think the second makes sense here.
BTW, it also has implication on the documentation review.

Author: Done, added to risks

    - Security: no new RBAC or authentication changes.

>     - Monitoring/Observability: no new metrics or alerts required. The existing VM network interface info metric should be validated to confirm it reflects the updated network attachment after the swap.

Reviewer 8: on one hand `metrics` is listed as not covered but here the statement is `The existing VM network interface info metric should be validated to confirm it reflects the updated network attachment after the swap.` and appears under tests scnearios - should be listed uner NFRs that are tested

Author: Fixed

#### **2. Known Limitations**

- **Non-migratable VMs:** Changing the secondary network on non-migratable VMs is not supported; the VM will not reconnect to the new network without a restart.
  - *Sign-off:* Ronen Sde-Or, 04/2026

> - **Binding support:** Any supported bridge binding only.

Reviewer: >  Any supported bridge binding only

I don't understand this sentence. Is this a typo?

Author: Edited, thanks

>     - Scalability: the feature has scale implications — large-scale NAD reference updates are throttled by the cluster's migration parallelism limits. This is accepted as a known risk rather than a test requirement (see Risks, Section II.5). Documentation should reflect this limitation.

Reviewer 8: not testing something is not a risk, it is out ot scope and should be placed under the oos section.

Author: Fixed

  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Binding support:** Secondary network reassignment is supported for bridge binding only; other binding types are not in scope for this feature.

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

- **Single-node clusters (SNO):** Cannot be supported because migration is not possible.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Migrating between CNI types:** Explicit non-goal.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Changing network binding/plugin:** Explicit non-goal.
  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Seamless network connectivity until action is completed:** Not guaranteed by design.
  - *Sign-off:* Ronen Sde-Or, 04/2026

> - **No limitation of migration retries because of missing Network Attachment Definition:** Explicit non-goal.

Reviewer 3: Why is this limited to missing NAD? Many reasons could exist for a migration failure.

Have you really convinced the PM that this is acceptable? Very odd.

Author: These limitations were accepted by Ronen on the previous version of the STP https://github.com/RedHatQE/openshift-virtualization-tests-design-docs/pull/39#discussion_r2939960932

Reviewer 3: > Why is this limited to missing NAD? Many reasons could exist for a migration failure.

I'm unclear if you answered me on this one.

Author: Fixed, now the risk refers to any migration failure

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

  - *Sign-off:* Ronen Sde-Or, 04/2026

- **Limiting migration retries on failure:** Not supported.

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

- [x] **Developer Handoff/QE Kickoff**
  - *Key takeaways and concerns:* Meeting where Dev introduced QE through NAD change in VM, migration condition handling, and compatibility with DNC.

- [x] **Technology Challenges**
  - *List identified challenges:* The feature relies on live migration, which is not supported on all cluster topologies (e.g. single-node clusters), limiting the environments in which this feature can be tested.
  - *Impact on testing approach:* Tests must run on multi-node clusters where live migration is available.

- [x] **API Extensions**
  - *List new or modified APIs:* No new APIs introduced by this feature.
  - *Testing impact:* N/A

- [x] **Test Environment Needs**
  - *See environment requirements in Section II.3 and testing tools in Section II.3.1*

- [x] **Topology Considerations**
  - *Describe topology requirements:* No known limitations when considering different topologies.
  - *Impact on test design:* No specific impact on test design beyond standard live migration requirements.

### **II. Software Test Plan (STP)**

#### **1. Scope of Testing**

> **Testing Goals**

Reviewer 3: Negative scenarios have not been covered per what I see, although they have been raised as limitations.
E.g. a non migratable VM scenario.

Author: Added non migratable VM scenario

> - **[P0]** Verify that a running VM with bridge binding can be reconnected to a different secondary network without rebooting, has connectivity on the new network, and is no longer reachable on the previous network.

Reviewer 3: This is not defined in an actionable manner. While the binding is defined, the CNI used is not.
The next line does mention indirectly the CNI and its configuration (localnet), which is fine.

Author: Added

Reviewer 3: You have somehow linked between bridge binding and the Linux bridge CNI as if it is trivial (the parenthesis expresses that). I do not understand.

Author: Rephrased

- **[P1]** Verify that a running VM connected to a localnet secondary network can be reconnected to a different localnet network without rebooting and has connectivity on the new network.

> - **[P1]** Verify that multiple secondary networks on the same running VM can each be independently updated and remain reachable.

Reviewer 3: If you add something like this, it only makes sense to me that you check it works if you change multiple networks at once. Independently is vague and can be part of the prev scenarios (not that you need to mention it).

Author: IIUC if we change N networks, we expect N sequential migrations.
So we can use the first test case to run in N times and not create a dedicated test case for this.
Did I understand you correctly?

Reviewer 3: > IIUC if we change N networks, we expect N sequential migrations.

If the patching was done in one shot, or very quickly, this is incorrect.

Author: So how can we test this feature on a VM with multiple secondary interfaces?

Reviewer 3: I do not understand what scenario you want to check and why it is important.
I understand that multiple interfaces can be changed is a thing.

What exactly are you after?

Reviewer 6: There could be a scenario where the user wants to change a VM with :
iface1->nadA, iface2->nadA, iface3->nadB, iface4->nadB
to 
iface1->nadA, iface2->nadB, iface3->nadA, iface3->nadB
in one go. 
If the change is done as a single patch it will happen in a single migration. 
We have a [UT](https://github.com/kubevirt/kubevirt/blob/1b33c989a8ab14027a4e47fbddd3a867ab990086/pkg/network/controllers/vm_test.go#L700) to check that our sync func handles this type of scenarios correctly. So theoretically we can skip an e2e test for this. But changing multiple NADs seem to be a real customer scenario, so maybe we should have an e2e test?

Author: IMO we need a scenario of changing VLAN of multiple interfaces in a single patch and test connectivity
I edited the test case accordingly

> - **[P2]** Verify that a VM can be successfully live-migrated after completing a NAD reference update, and maintains connectivity on the new network.

Reviewer: To be clear, the NAD change triggers a migration, and in this test we will perform a second migration correct?

Author: Yes, edited for clarification

> - **[P2]** Verify that a VM can be successfully live-migrated (user-triggered) after completing a NAD reference update, and maintains connectivity on the new network.

Reviewer 3: I do not see why this is needed.
Also, there is no such thing as user-triggered. Only cluster-admin triggered, which does not sound relevant.
Please elaborate why this scenario is required.

Author: The main concern: does the feature's internal migration leave the VM in a clean state for subsequent admin-triggered migrations?

Reviewer 3: Once the VM migrated and the state of the VM stabilized, I do not see how the system will identify that it reached that state from a hotplug or not. What mechanics are playing here which caused you to consider this scenario as risky enough to cover?

Author: We do test connectivity post migration on a VM with localnet setup so I thought we should test it here as well.

> Once the VM migrated and the state of the VM stabilized, I do not see how the system will identify that it reached that state from a hotplug or not

Now that I know that, it does sound redundant, removed, thanks.

> - **[P2]** Verify that a VM with an updated secondary network reference maintains connectivity on the new network after cluster upgrade.

Reviewer 3: Same, I see no logic for this scenario.
Once the configuration has been updated and it reached a stable state (i.e. it changed the network), the system does not know to differentiate anymore between a new VM created on the new network or an older VM that was changed to the new network.

Author: Removed, thanks

- **[P0]** Verify that a running VM with bridge binding on a Linux bridge CNI secondary network can be reconnected to a different secondary network without rebooting, has connectivity on the new network, and is no longer reachable on the previous network.
- **[P1]** Verify that a running VM connected to a localnet secondary network can be reconnected to a different localnet network without rebooting and has connectivity on the new network.
- **[P1]** Verify that when NAD references for multiple bridge binding on Linux bridge CNI secondary network interfaces on the same running VM are updated at once, all interfaces establish connectivity on their respective new networks.

> - **[P1]** Verify that attempting to change the secondary network VLAN on a non-migratable VM with bridge binding on a Linux bridge CNI secondary network does not succeed, and the VM remains on the original network.

Reviewer 8: should the tests also cover modification of non-gridge bindings? between cni types?

Author: In theory we could have a negative scenario with passt binding but it's not available on 4.22

- [x] **Regression Testing** — Verifies that new changes do not break existing functionality
  - *Details:* Existing secondary network and live migration test suites will run as regression to confirm no impact.

**Non-Functional**

- [ ] **Performance Testing** — Validates feature performance meets requirements (latency, throughput, resource usage)
  - *Details:* Not applicable; no new performance requirements introduced for this feature.

- [ ] **Scale Testing** — Validates feature behavior under increased load and at production-like scale
  - *Details:* Not applicable; no new scale requirements introduced for this feature.

- [ ] **Security Testing** — Verifies security requirements, RBAC, authentication, authorization
  - *Details:* Not applicable; no new RBAC or authentication changes introduced.

- [ ] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g. status conditions, clear output)

>   - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g. status conditions, events, or notifications indicating success or failure)

Reviewer 6: [not a review comment - just an fyi]: The only way right now to know that the NAD hotplug completed is to monitor the conditions (like migrationRequired appears, then disappears, GAConnected reappears). So the UX might not be very smooth.

Author: Will it be tested in Tier1?

Reviewer 9: @azhivovk the UI tests for network are in tier2

Reviewer 9: @frenzyfriday yes, UI test should verify it using API

  - *Details:* Not applicable; no new RBAC or authentication changes introduced.

- [x] **Usability Testing** — Validates user experience and accessibility requirements
  - Does the feature require a UI? If so, ensure the UI aligns with the requirements (UI/UX consistency, accessibility)
  - Does the feature expose CLI commands? If so, validate usability and that needed information is available (e.g. status conditions, clear output)
  - Does the feature trigger backend operations that should be reported to the admin? If so, validate that the user receives clear feedback about the operation and its outcome (e.g. status conditions, events, or notifications indicating success or failure)

>   - *Details:* QE UI team will validate that the VM network display reflects the updated network name after the swap, and that the previously assigned IP is preserved in the display.

Reviewer 8: please add a jira ref

Author: Done

  - Does the feature maintain backward compatibility with previous API versions and configurations?
  - *Details:* Not applicable; no new API versions or configurations introduced.

- [x] **Upgrade Testing** — Validates upgrade paths from previous versions, data migration, and configuration preservation
  - *Details:* Verify that VMs with updated NAD references retain their network configuration and connectivity after cluster upgrade.

- [ ] **Dependencies** — Blocked by deliverables from other components/products. Identify what we need from other teams before we can test.
  - *Details:* Not applicable; no blocking external dependencies identified.

- [ ] **Cross Integrations** — Does the feature affect other features or require testing by other teams? Identify the impact we cause.
  - *Details:* Not applicable; no impact on other features or teams identified.

**Infrastructure**

- [ ] **Cloud Testing** — Does the feature require multi-cloud platform testing?
  - *Details:* Not applicable; feature requires bare-metal cluster with live migration support.

#### **3. Test Environment**

> - **Cluster Topology:** 3-master/3-worker bare-metal (minimum 2 workers), multi NIC nodes

Author: Done

- **OCP & OpenShift Virtualization Version(s):** OCP 4.22 with OpenShift Virtualization version 4.22

- **CPU Virtualization:** VT-x (Intel) or AMD-V enabled or ARM

- **Compute Resources:** Minimum per worker node: 8 vCPUs, 32GB RAM

- **Special Hardware:** N/A

- **Storage:** ocs-storagecluster-ceph-rbd-virtualization

- **Network:** OVN-Kubernetes, IPv4, IPv6, secondary bridge network, localnet

- **Required Operators:** NMState Operator (for bridge interface configuration)

> - **Platform:** Agnostic

Author: We could run these tests on any cluster plastform, not only bare metal

> - **Platform:** Agnostic

Reviewer 5: @azhivovk, I think the bot has a point here.
If we cannot run the tests on a cloud for example, they are not agnostic.

> - **Platform:** Agnostic

Author: @coderabbitai changed to Bare metal, PSI

- **Special Configurations:** N/A

#### **3.1. Testing Tools & Frameworks**

- **Test Framework:** Standard

- **CI/CD:** N/A

- **Other Tools:** N/A

#### **4. Entry Criteria**

The following conditions must be met before testing can begin:

- [x] Requirements and design documents are **approved and merged**
- [x] Test environment can be **set up and configured** (see Section II.3 - Test Environment)

#### **5. Risks**

**Timeline/Schedule**

- [x] **Risk:** None identified.
  - **Mitigation:** N/A
  - *Estimated impact on schedule:* N/A
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Test Coverage**

- [x] **Risk:** Coverage is limited to bridge binding only; other binding types are not tested.
  - **Mitigation:** Accepted per feature scope - bridge binding is the only supported binding type for this feature.
  - *Areas with reduced coverage:* Other network binding types.
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Test Environment**

> - [x] **Risk:** Live migration requires at least 2 worker nodes; SNO clusters cannot be used.

Reviewer: silly question but why is this a risk?

Author: You are right, removed
Thanks

  - **Mitigation:** Ensure test environment has a minimum of 2 worker nodes as specified in Section II.3.
  - *Missing resources or infrastructure:* None identified at this time.
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Untestable Aspects**

> - [x] **Risk:** Seamless network connectivity during the NAD swap is not guaranteed by design and cannot be tested as a pass/fail condition.

Reviewer: It's a non-goal therefore it's not an untestable aspect.

Author: Done

Reviewer: I think that you can drop it completely.

Author: According to the [template](https://github.com/RedHatQE/openshift-virtualization-tests-design-docs/blob/main/stps/stp-template/stp.md?plain=1#L344) we need to keep it and explain why there's no risk

```

#### **5. Risks**

<!-- Document specific risks for this feature. If a risk category is not applicable, mark as "N/A" with
justification in mitigation strategy. -->
```

  - **Mitigation:** Tests will verify connectivity before and after the swap; transient disruption during swap is accepted.
  - *Alternative validation approach:* Verify the VM has connectivity on the new network after the swap completes, regardless of whether prior connections were preserved.
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Resource Constraints**

- [x] **Risk:** No special resource requirements.
  - **Mitigation:** N/A
  - *Current capacity gaps:* None
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Dependencies**

- [x] **Risk:** No blocking external dependencies identified.
  - **Mitigation:** N/A
  - *Dependent teams or components:* None
  - *Sign-off:* Ronen Sde-Or, 04/2026

**Other**

- [x] **Risk:** None identified.
  - **Mitigation:** N/A
  - *Sign-off:* Ronen Sde-Or, 04/2026

---

### **III. Test Scenarios & Traceability**

> - **[VIRTSTRAT-560]** — As a VM owner, I want to re-assign my VM to a different secondary network without rebooting so that my workload remains available during network reconfiguration.

Reviewer: See my comment above for the objective o the usecase.

Author: Done

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 8: this is not a risk. A risk is something that is uncertain, might impact testing, it is something that you took an decision not to test. 
not testing large-scale NAD reference updates should be under out of scope with the explanation why it is not tested.

Author: Moved to OOS
But it's affecting test coverage
Maybe it should be both OOS and a risk?

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 2: Approved

>   - *Sign-off:* Ronen Sde-Or, 04/2026

Reviewer 8: this is not a risk; if the feature does not support other bindings, it is a limitation. 
a risk is something we plan on doing but may be affected by ....

Author: Removed
